software/front/US_BootLoader.X/io.h

```diff
@@ -49,17 +62,15 @@ typedef union {
     };
 }USBPacket;
 
-
-IOC_EXTERN volatile USBPacket RXP;
-IOC_EXTERN volatile USBPacket TXP;
+/*Buffers used for USB communication*/
+IOC_EXTERN USBPacket RXP;
+IOC_EXTERN USBPacket TXP;
```

Comment:
The buffers used for USB communication should not be declared as volatile. The buffer is not modified in any way by the HW so it seems by using volatile is leading to strange things.

Reply:
From the XC32 manual:
..and any variables which may be modified by interrupt routines should use
this qualifier as well. For example:

USB routines can run in interrupt mode (although at the moment it does not). And I don't know if these buffers are modified by hardware (DMA or something else) or not.

The USB stack used is not only for this project - I may use it in the future (if not already, I don't remember) for another projects in interrupt mode.

Reply:
Well I have not yet found the reason why this is not working with volatile.
Your define was actually using the buffer USBTxBuffer 
#define TXP (*((USBPacket *)USBTxBuffer))

which is declared in usb_driver.c as unsigned char USBRxBuffer[64]

Reply:
Also:
/*memory move necessary because the command is on 1 byte, could be avoided by using 4 bytes*/
Yes, it can be avoided, but we will lose 3 bytes of data and the PC software must be edited to support this. The bootloader uses USB HID, and the pockets there are 64 bytes if I remember well.

I will think of a better way to do this somehow.

Reply:
Yes, I've looked on it too, that's why I only added the comment just to remember why that memmove is necessary.
The PC app I've seen and you have used just binary array to fill in the data, as it looks there are too many changes to make the command on 4 bytes